File: langs/swift/Sources/VMx/State/DiscriminatorVM.swift
//
// DiscriminatorVM — owns one active key with modal precedence helpers.
//
import Combine

public final class DiscriminatorVM<Key: Equatable> {
    private var activeChangedSubject = PassthroughSubject<Key, Never>()
    private var modalStack: [Key] = []
    private var disposed = false

    public private(set) var activeKey: Key

    public init(initial: Key) {
        self.activeKey = initial
    }

    public var activeChanged: AnyPublisher<Key, Never> {
        activeChangedSubject.eraseToAnyPublisher()
    }

    public var modalDepth: Int {
        modalStack.count
    }

    public func isActive(_ key: Key) -> Bool {
        activeKey == key
    }

    public func setActiveKey(_ key: Key) {
        guard !disposed else { return }
        modalStack.removeAll()
        setActiveKeyPreservingModals(key)
    }

    private func setActiveKeyPreservingModals(_ key: Key) {
        guard key != activeKey else { return }
        activeKey = key
        activeChangedSubject.send(key)
    }

    public func modalOpen(_ modalKey: Key) {
        guard !disposed else { return }
        modalStack.append(activeKey)
        setActiveKeyPreservingModals(modalKey)
    }

    public func modalClose() {
        guard !disposed, let previous = modalStack.popLast() else { return }
        setActiveKeyPreservingModals(previous)
    }

    public func clearModals() {
        guard !disposed else { return }
        modalStack.removeAll()
    }

    public func dispose() {
        guard !disposed else { return }
        disposed = true
        modalStack.removeAll()
        activeChangedSubject.send(completion: .finished)
    }
}

File: langs/typescript/src/state/discriminatorVM.ts
/**
 * DiscriminatorVM — owns one active key with modal precedence helpers.
 */
import { Observable, Subject } from "rxjs";

export class DiscriminatorVM<TKey> {
  #activeKey: TKey;
  readonly #modalStack: TKey[] = [];
  readonly #activeChanged = new Subject<TKey>();
  #disposed = false;

  constructor(initial: TKey) {
    this.#activeKey = initial;
  }

  get activeKey(): TKey {
    return this.#activeKey;
  }

  get activeChanged(): Observable<TKey> {
    return this.#activeChanged.asObservable();
  }

  get modalDepth(): number {
    return this.#modalStack.length;
  }

  isActive(key: TKey): boolean {
    return Object.is(this.#activeKey, key) || this.#activeKey === key;
  }

  setActiveKey(key: TKey): void {
    if (this.#disposed) return;
    this.#modalStack.length = 0;
    this.#setActiveKeyPreservingModals(key);
  }

  #setActiveKeyPreservingModals(key: TKey): void {
    if (this.isActive(key)) return;
    this.#activeKey = key;
    this.#activeChanged.next(key);
  }

  modalOpen(modalKey: TKey): void {
    if (this.#disposed) return;
    this.#modalStack.push(this.#activeKey);
    this.#setActiveKeyPreservingModals(modalKey);
  }

  modalClose(): void {
    if (this.#disposed || this.#modalStack.length === 0) return;
    const previous = this.#modalStack.pop() as TKey;
    this.#setActiveKeyPreservingModals(previous);
  }

  clearModals(): void {
    if (this.#disposed) return;
    this.#modalStack.length = 0;
  }

  dispose(): void {
    if (this.#disposed) return;
    this.#disposed = true;
    this.#modalStack.length = 0;
    this.#activeChanged.complete();
  }
}
